src/library/scala/util/Using.scala

```diff
@@ -40,66 +43,85 @@ import scala.util.control.{ControlThrowable, NonFatal}
   *
   * @example
   * {{{
-  * val lines: Try[Seq[String]] = for {
-  *   r1 <- Using(resource1)
-  *   r2 <- Using(resource2)
-  *   r3 <- Using(resource3)
-  *   r4 <- Using(resource4)
-  * } yield {
-  *   // use your resources here
-  *   r1.lines ++ r2.lines ++ r3.lines ++ r4.lines
-  * }
+  * val lines: Try[Seq[String]] = {
+  *   for {
+  *     r1 <- Using(resource1)
+  *     r2 <- Using(resource2)
+  *     r3 <- Using(resource3)
+  *     r4 <- Using(resource4)
+  *   } yield {
+  *     // use your resources here
+  *     r1.lines ++ r2.lines ++ r3.lines ++ r4.lines
+  *   }
+  * }.use()
   * }}}
   */
-final class Using[R] private(resource: => R) {
-  private[this] val used = new AtomicBoolean(false)
+final class Using[R, A] private(resource: Using.SingleUse[R], op: R => A) {
+  import Using._
 
-  /** Performs an operation using a resource, and then releases the resource,
+  /** Performs an operation on the result of this `Using`'s existing operation
+    * using its resource, and then releases the resource,
     * even if the operation throws an exception.
     *
     * @param f the operation to perform
     * @param r an implicit [[Using.Resource]]
-    * @tparam A the return type of the operation
+    * @tparam B the return type of the operation
     * @throws java.lang.IllegalStateException if the resource has already been used
     * @return a [[scala.util.Try `Try`]] containing the result of the operation, or
     *         an exception if one was thrown by the operation or by releasing the resource
     */
   @throws[IllegalStateException]("if the resource has already been used")
-  @inline def apply[A](f: R => A)(implicit r: Using.Resource[R]): Try[A] = map(f)
+  def apply[B](f: A => B)(implicit r: Resource[R]): Try[B] = map(f).use()
 
-  /** Performs an operation using a resource, and then releases the resource,
+  /** Returns a `Using` which will perform the specified operation (after this `Using`'s
+    * operation) when [[use() used]].
+    *
+    * @param f the operation to perform
+    * @tparam B the return type of the operation
+    */
+  def map[B](f: A => B): Using[R, B] =
+    new Using(resource, if (op eq cachedIdentity) f.asInstanceOf[R => B] else r => f(op(r)))
+
+  /** Returns a `Using` which will perform the specified `Using`-returning operation
+    * and [[use() use]] the returned `Using` (after this `Using`'s operation) when [[use() used]].
+    *
+    * @param f the `Using`-returning operation to perform
+    * @tparam B the return type of the returned `Using`'s operation(s)
+    */
+  def flatMap[R1: Resource, B](f: A => Using[R1, B]): Using[R, B] =
+    new Using(resource, r => f(op(r)).use0())
+
+  def filter(p: A => Boolean): Using[R, A] =
+    new Using(resource, r => {
+      val res = op(r)
+      if (p(res)) res
+      else throw new NoSuchElementException("Predicate does not hold for " + res)
+    })
```

Comment:
This is not really related to this PR, but is it a good idea to have filter if it throws an exception? I think we should not have it.

Reply:
It seems relatively related to this PR to me.

The question is, if not `NoSuchElementException`, what does the eventual `Try` contain?

(Side note: I tried out an optimisation where it uses a cached `ControlThrowable` initially, and then only turns it into a `NoSuchElementException` if you ask for a `Try` back. If you call `foreach`, it just catches the cached throwable and does nothing. Other than being a bit confusing from a maintenance perspective, I didn't see any downsides.)

Reply:
> The question is, if not `NoSuchElementException`, what does the eventual `Try` contain?

Any exception that can be thrown here: https://github.com/scala/scala/blob/56c45cd8d65e13af688892fc3831bbba9b8a8a28/src/library/scala/util/Using.scala#L181-L217

Reply:
I'm fine with removing `filter`. If we decide it's a good idea later, it can always be added back.